### PR TITLE
Convert Http2Reader, Http2Writer to Kotlin

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/internal/http2/ErrorCode.kt [hunk 1]
```diff
@@ -40,16 +40,12 @@
...
-
-  ErrorCode(int httpCode) {
-    this.httpCode = httpCode;
-  }
-
-  public static ErrorCode fromHttp2(int code) {
-    for (ErrorCode errorCode : ErrorCode.values()) {
-      if (errorCode.httpCode == code) return errorCode;
+  companion object {
+    fun fromHttp2(code: Int): ErrorCode? {
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/http2/ErrorCode.kt]:
```
    fun fromHttp2(code: Int): ErrorCode? = values().find { it.httpCode == code }
```

File: okhttp/src/main/java/okhttp3/internal/http2/Http2.kt [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+      "DATA", "HEADERS", "PRIORITY", "RST_STREAM", "SETTINGS", "PUSH_PROMISE", "PING", "GOAWAY",
+      "WINDOW_UPDATE", "CONTINUATION"
+  )
+
+  /**
+   * Lookup table for valid flags for DATA, HEADERS, CONTINUATION. Invalid combinations are
+   * represented in binary.
+   */
+  private val FLAGS = arrayOfNulls<String>(0x40) // Highest bit flag is 0x20.
+  private val BINARY = arrayOfNulls<String>(256)
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2.kt]:
```
  private val BINARY = Array(256) {
    Integer.toBinaryString(it).let { s -> s.padStart(8 - s.length, '0') }
  }
```

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Reader.kt [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+  fun readConnectionPreface(handler: Handler) {
+    if (client) {
+      // The client reads the initial SETTINGS frame.
+      if (!nextFrame(true, handler)) {
+        throw IOException("Required SETTINGS preface not received")
+      }
+    } else {
+      // The server reads the CONNECTION_PREFACE byte string.
+      val connectionPreface = source.readByteString(Http2.CONNECTION_PREFACE.size.toLong())
+      if (logger.isLoggable(FINE)) logger.fine(format("<< CONNECTION %s", connectionPreface.hex()))
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2Reader.kt]: minor simplification

```
      if (logger.isLoggable(FINE)) logger.fine("<< CONNECTION ${connectionPreface.hex()}")
```

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.java [hunk 4]
```diff
@@ -166,7 +166,8 @@
     // Like newSingleThreadExecutor, except lazy creates the thread.
     pushExecutor = new ThreadPoolExecutor(0, 1, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<>(),
         Util.threadFactory(Util.format("OkHttp %s Push Observer", connectionName), true));
-    peerSettings.set(Settings.INITIAL_WINDOW_SIZE, DEFAULT_INITIAL_WINDOW_SIZE);
+    peerSettings.set(Settings.INITIAL_WINDOW_SIZE,
+        DEFAULT_INITIAL_WINDOW_SIZE);
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.java]: Revert?

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Reader.kt [hunk 5]
```diff
@@ -0,0 +1,506 @@
...
+
+    return true
+  }
+
+  @Throws(IOException::class)
+  private fun readHeaders(handler: Handler, length: Int, flags: Int, streamId: Int) {
+    if (streamId == 0) throw IOException("PROTOCOL_ERROR: TYPE_HEADERS streamId == 0")
+
+    val endStream = (flags and FLAG_END_STREAM) != 0
+    val padding = if (flags and FLAG_PADDED != 0) source.readByte() and 0xff else 0
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2Reader.kt]: Is the conditional's precedence correct here? Above you parenthesize the `and` operation before the not-equals check.

Turns out there's a lot of unparenthesized versions below as well, so I assume yes.

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Reader.kt [hunk 6]
```diff
@@ -0,0 +1,506 @@
...
+    if (flags and FLAG_ACK != 0) {
+      if (length != 0) throw IOException("FRAME_SIZE_ERROR ack frame should be empty!")
+      handler.ackSettings()
+      return
+    }
+
+    if (length % 6 != 0) throw IOException("TYPE_SETTINGS length % 6 != 0: $length")
+    val settings = Settings()
+    var i = 0
+    while (i < length) {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2Reader.kt]:
```kotlin
for (i in 0 until length stepBy 6)
```

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Writer.kt [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+  private var maxFrameSize: Int = INITIAL_MAX_FRAME_SIZE
+  private var closed: Boolean = false
+  val hpackWriter: Hpack.Writer = Hpack.Writer(hpackBuffer)
+
+  @Synchronized @Throws(IOException::class)
+  fun connectionPreface() {
+    if (closed) throw IOException("closed")
+    if (!client) return // Nothing to write; servers don't send connection headers!
+    if (logger.isLoggable(FINE)) {
+      logger.fine(format(">> CONNECTION %s", CONNECTION_PREFACE.hex()))
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2Writer.kt]: We should make a little DSL ditty for this pattern in a follow-up.
```kotlin
fun Logger.fine(body: () -> String) {
  if (isLoggable(FINE)) find(body())
}
```

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Writer.kt [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+    }
+    sink.write(CONNECTION_PREFACE.toByteArray())
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2Writer.kt]: unrelated to conversion: `toByteArray()` shouldn't be needed? I can do a pass of this over the entire codebase once merged.

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Writer.kt [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+  fun applyAndAckSettings(peerSettings: Settings) {
+    if (closed) throw IOException("closed")
+    this.maxFrameSize = peerSettings.getMaxFrameSize(maxFrameSize)
+    if (peerSettings.headerTableSize != -1) {
+      hpackWriter.setHeaderTableSizeSetting(peerSettings.headerTableSize)
+    }
+    val length = 0
+    val type = TYPE_SETTINGS
+    val flags = FLAG_ACK
+    val streamId = 0
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2Writer.kt]: Inline these 4

Seems like a pattern in the file. But seems like it's just wanting for named arguments.

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Writer.kt (hunk 7 above)
Comment (yschimke) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Writer.kt]: probably inline as well?

Comment (JakeWharton) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Writer.kt]: Whoops, yep. Definitely `inline`!

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Reader.kt [hunk 10]
```diff
@@ -0,0 +1,506 @@
...
+    // |                 Length (24)                   |
+    // +---------------+---------------+---------------+
+    // |   Type (8)    |   Flags (8)   |
+    // +-+-+-----------+---------------+-------------------------------+
+    // |R|                 Stream Identifier (31)                      |
+    // +=+=============================================================+
+    // |                   Frame Payload (0...)                      ...
+    // +---------------------------------------------------------------+
+    val length = source.readMedium()
+    if (length < 0 || length > INITIAL_MAX_FRAME_SIZE) {
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2Reader.kt]: Is negative possible with readMedium?

File: okhttp/src/main/java/okhttp3/internal/http2/ErrorCode.kt (hunk 1 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/ErrorCode.kt]: Fixed.

File: okhttp/src/main/java/okhttp3/internal/http2/Http2.kt (hunk 2 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2.kt]: Fixed!

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Reader.kt (hunk 3 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Reader.kt]: Fixed!

File: okhttp/src/main/java/okhttp3/internal/UtilKt.kt [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+import okio.BufferedSink
+import okio.BufferedSource
+import java.io.IOException
+
+infix fun Byte.and(mask: Int): Int = toInt() and mask
+infix fun Short.and(mask: Int): Int = toInt() and mask
+infix fun Int.and(mask: Long): Long = toLong() and mask
+
+@Throws(IOException::class)
+fun BufferedSink.writeMedium(medium: Int) {
```
[the remark this answers is not in the document]
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/UtilKt.kt]: Fixed!

[the remark this answers is not in the document]
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/UtilKt.kt]: (just kidding!)

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.java (hunk 4 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.java]: Fixed.

File: okhttp/src/main/java/okhttp3/internal/http2/Settings.kt [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+  val initialWindowSize: Int
+    get() {
+      val bit = 1 shl INITIAL_WINDOW_SIZE
+      return if (bit and set != 0) values[INITIAL_WINDOW_SIZE] else DEFAULT_INITIAL_WINDOW_SIZE
+    }
+
+  fun clear() {
+    set = 0
+    Arrays.fill(values, 0)
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/http2/Settings.kt]:
```
    values.fill(0)
```

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Reader.kt (hunk 5 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Reader.kt]: No action taken!

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Reader.kt (hunk 6 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Reader.kt]: Fixed!

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Writer.kt (hunk 7 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Writer.kt]: Yep. Follow-up!

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Writer.kt (hunk 8 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Writer.kt]: oooh, nice. Fixed.

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Writer.kt (hunk 9 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Writer.kt]: Fixed!

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Reader.kt (hunk 10 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Reader.kt]: Nope! Removed the dead case.

File: okhttp/src/main/java/okhttp3/internal/http2/Settings.kt (hunk 12 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Settings.kt]: OOooh fixed.

**Pull request merged**